src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(currentState, stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(currentState, stateCopy, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(currentState, stateCopy);
+        break;
+      default:
+        break;
+    }
+  }
+
+  return currentState;
+
+  function addProperties(newState, oldState, properties) {
+    Object.assign(oldState, properties);
+
+    const newStateObject = Object.assign({}, oldState);
+
+    newState.push(newStateObject);
+  }
+
+  function removeProperties(newState, oldState, removeKeys) {
+    for (const key of removeKeys) {
+      delete oldState[key];
+    }
+
+    const newStateObject = Object.assign({}, oldState);
+
+    newState.push(newStateObject);
```

Comment:
The `removeProperties` function directly mutates the `oldState` object by deleting keys. To comply with the checklist, you should create a new copy of `oldState` and then remove the keys from this copy, not from `oldState` directly.